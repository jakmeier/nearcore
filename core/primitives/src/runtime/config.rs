//! Settings of the parameters of the runtime.
use std::collections::BTreeMap;

use near_primitives_core::parameter::{load_parameters_from_txt, read_parameter, Parameter};
use serde::{Deserialize, Serialize};

use crate::config::VMConfig;
use crate::runtime::fees::RuntimeFeesConfig;
use crate::serialize::u128_dec_format;
use crate::types::{AccountId, Balance};

/// The structure that holds the parameters of the runtime, mostly economics.
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct RuntimeConfig {
    /// Amount of yN per byte required to have on the account.  See
    /// <https://nomicon.io/Economics/README.html#state-stake> for details.
    #[serde(with = "u128_dec_format")]
    pub storage_amount_per_byte: Balance,
    /// Costs of different actions that need to be performed when sending and processing transaction
    /// and receipts.
    pub transaction_costs: RuntimeFeesConfig,
    /// Config of wasm operations.
    pub wasm_config: VMConfig,
    /// Config that defines rules for account creation.
    pub account_creation_config: AccountCreationConfig,
}

impl RuntimeConfig {
    pub fn from_parameters(params: &BTreeMap<Parameter, String>) -> Self {
        Self {
            storage_amount_per_byte: read_parameter(&params, Parameter::StorageAmountPerByte),
            transaction_costs: RuntimeFeesConfig::from_parameters(&params),
            wasm_config: VMConfig::from_parameters(&params),
            account_creation_config: AccountCreationConfig::from_parameters(&params),
        }
    }

    pub fn from_parameters_txt(txt_file: &[u8]) -> Self {
        Self::from_parameters(&load_parameters_from_txt(txt_file))
    }

    pub fn test() -> Self {
        RuntimeConfig {
            // See https://nomicon.io/Economics/README.html#general-variables for how it was calculated.
            storage_amount_per_byte: 909 * 100_000_000_000_000_000,
            transaction_costs: RuntimeFeesConfig::test(),
            wasm_config: VMConfig::test(),
            account_creation_config: AccountCreationConfig::default(),
        }
    }

    pub fn free() -> Self {
        Self {
            storage_amount_per_byte: 0,
            transaction_costs: RuntimeFeesConfig::free(),
            wasm_config: VMConfig::free(),
            account_creation_config: AccountCreationConfig::default(),
        }
    }
}

/// The structure describes configuration for creation of new accounts.
#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq)]
pub struct AccountCreationConfig {
    /// The minimum length of the top-level account ID that is allowed to be created by any account.
    pub min_allowed_top_level_account_length: u8,
    /// The account ID of the account registrar. This account ID allowed to create top-level
    /// accounts of any valid length.
    pub registrar_account_id: AccountId,
}

impl Default for AccountCreationConfig {
    fn default() -> Self {
        Self {
            min_allowed_top_level_account_length: 0,
            registrar_account_id: "registrar".parse().unwrap(),
        }
    }
}
impl AccountCreationConfig {
    fn from_parameters(params: &BTreeMap<Parameter, String>) -> Self {
        Self {
            min_allowed_top_level_account_length: read_parameter(
                params,
                Parameter::MinAllowedTopLevelAccountLength,
            ),
            registrar_account_id: read_parameter(params, Parameter::RegistrarAccountId),
        }
    }
}
