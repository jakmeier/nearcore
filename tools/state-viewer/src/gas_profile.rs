//! State viewer functions to read gas profile information from execution
//! outcomes stored in RocksDB.

use near_primitives::profile::Cost;
use near_primitives::transaction::ExecutionOutcome;
use near_primitives_core::parameter::Parameter;
use node_runtime::config::RuntimeConfig;
use std::collections::BTreeMap;

pub(crate) struct GasFeeCounters {
    counters: BTreeMap<Parameter, u64>,
}

pub(crate) fn extract_gas_counters(
    outcome: &ExecutionOutcome,
    runtime_config: &RuntimeConfig,
) -> Option<GasFeeCounters> {
    match &outcome.metadata {
        near_primitives::transaction::ExecutionMetadata::V1 => None,
        near_primitives::transaction::ExecutionMetadata::V2(meta_data) => {
            let mut counters = BTreeMap::new();

            for param in Parameter::ext_costs() {
                match param.cost().unwrap_or_else(|| panic!("ext cost {param} must have a cost")) {
                    Cost::ExtCost { ext_cost_kind } => {
                        let parameter_value =
                            ext_cost_kind.value(&runtime_config.wasm_config.ext_costs);
                        let gas = meta_data.get_ext_cost(ext_cost_kind);
                        if parameter_value != 0 && gas != 0 {
                            assert_eq!(
                                0,
                                gas % parameter_value,
                                "invalid gas profile for given config"
                            );
                            let counter = gas / parameter_value;
                            *counters.entry(*param).or_default() += counter;
                        }
                    }
                    _ => unreachable!("{param} must be ExtCost"),
                };
            }

            let num_wasm_ops = meta_data[Cost::WasmInstruction]
                / runtime_config.wasm_config.regular_op_cost as u64;
            if num_wasm_ops != 0 {
                *counters.entry(Parameter::WasmRegularOpCost).or_default() += num_wasm_ops;
            }

            // TODO: Action costs should also be included.
            // This is tricky, however. From just the gas numbers in the profile
            // we cannot know the cost is split to parameters. Because base and byte
            // costs are all merged. Same for different type of access keys.
            // The only viable way right now is go through each action separately and
            // recompute the gas cost from scratch. For promises in function
            // calls that includes looping through outgoing promises and again
            // recomputing the gas costs.
            // And of course one has to consider that some actions will be SIR
            // and some will not be.
            //
            // For now it is not clear if implementing this is even worth it.
            // Alternatively, we could also make the profile data more detailed.

            Some(GasFeeCounters { counters })
        }
    }
}

impl std::fmt::Display for GasFeeCounters {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        for (param, counter) in self.counters.iter() {
            writeln!(f, "{param:<48} {counter:>16}")?;
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use near_primitives::profile::ProfileData;
    use near_primitives::transaction::ExecutionMetadata;
    use near_primitives::types::AccountId;
    use node_runtime::config::RuntimeConfig;

    #[test]
    fn test_extract_gas_counters() {
        let config = RuntimeConfig::test();
        let costs = [
            (Parameter::WasmStorageWriteBase, 137),
            (Parameter::WasmStorageWriteKeyByte, 4629),
            (Parameter::WasmStorageWriteValueByte, 2246),
            // note: actions are not included in profile, yet
            (Parameter::ActionDeployContractExecution, 2 * 184765750000),
            (Parameter::ActionDeployContractSendSir, 2 * 184765750000),
            (Parameter::ActionDeployContractPerByteSendSir, 1024 * 6812999),
            (Parameter::ActionDeployContractPerByteExecution, 1024 * 64572944),
            (Parameter::WasmRegularOpCost, 7000),
        ];

        let outcome = create_execution_outcome(&costs, &config);
        let profile = extract_gas_counters(&outcome, &config).expect("no counters returned");

        insta::assert_display_snapshot!(profile);
    }

    fn create_execution_outcome(
        costs: &[(Parameter, u64)],
        config: &RuntimeConfig,
    ) -> ExecutionOutcome {
        let mut gas_burnt = 0;
        let mut profile_data = ProfileData::new();
        for &(parameter, value) in costs {
            match parameter.cost() {
                Some(Cost::ExtCost { ext_cost_kind }) => {
                    let gas = value * ext_cost_kind.value(&config.wasm_config.ext_costs);
                    profile_data.add_ext_cost(ext_cost_kind, gas);
                    gas_burnt += gas;
                }
                Some(Cost::WasmInstruction) => {
                    let gas = value * config.wasm_config.regular_op_cost as u64;
                    profile_data[Cost::WasmInstruction] += gas;
                    gas_burnt += gas;
                }
                // Multiplying for actions isn't possible because costs can be
                // split into multiple parameters. Caller has to specify exact
                // values.
                Some(Cost::ActionCost { action_cost_kind }) => {
                    profile_data.add_action_cost(action_cost_kind, value);
                    gas_burnt += value;
                }
                _ => unimplemented!(),
            }
        }
        let metadata = ExecutionMetadata::V2(profile_data);
        let account_id: AccountId = "alice.near".to_owned().try_into().unwrap();
        ExecutionOutcome {
            logs: vec![],
            receipt_ids: vec![],
            gas_burnt,
            tokens_burnt: 0,
            executor_id: account_id.clone(),
            status: near_primitives::transaction::ExecutionStatus::SuccessValue(vec![]),
            metadata,
        }
    }
}
