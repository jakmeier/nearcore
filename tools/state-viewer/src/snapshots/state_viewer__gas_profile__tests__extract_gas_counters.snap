---
source: tools/state-viewer/src/gas_profile.rs
expression: profile
---
wasm_regular_op_cost                                         7000
wasm_storage_write_base                                       137
wasm_storage_write_key_byte                                  4629
wasm_storage_write_value_byte                                2246

